src/tls.rs

```diff
@@ -256,6 +261,137 @@ fn open_read(filename: &Path, note: &'static str) -> crate::Result<Vec<u8>> {
     Ok(text)
 }
 
+pub struct MaybeTlsIncoming<I: Stream> {
+    incoming: I,
+    acceptor: Option<tokio_tls::TlsAcceptor>,
+    state: MaybeTlsIncomingState<I::Item>,
+}
+
+enum MaybeTlsIncomingState<S> {
+    Inner,
+    Accepting(tokio_tls::Accept<S>),
+}
+
+impl<I: Stream> MaybeTlsIncoming<I> {
+    pub fn new(incoming: I, tls: Option<TlsSettings>) -> crate::Result<Self> {
+        let acceptor = if let Some(tls) = tls {
+            let acceptor = tls.acceptor()?;
+            Some(acceptor.into())
+        } else {
+            None
+        };
+
+        let state = MaybeTlsIncomingState::Inner;
+
+        Ok(Self {
+            incoming,
+            acceptor,
+            state,
+        })
+    }
+}
+
+impl MaybeTlsIncoming<Incoming> {
+    pub fn bind(addr: &SocketAddr, tls: Option<TlsSettings>) -> crate::Result<Self> {
+        let listener = TcpListener::bind(addr)?;
+        let incoming = listener.incoming();
+
+        MaybeTlsIncoming::new(incoming, tls)
+    }
+}
+
+impl<I> Stream for MaybeTlsIncoming<I>
+where
+    I: Stream,
+    I::Item: AsyncRead + AsyncWrite,
+    I::Error: Into<crate::Error>,
+{
+    type Item = MaybeTlsStream<I::Item>;
+    type Error = crate::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            match &mut self.state {
+                MaybeTlsIncomingState::Inner => {
+                    let stream = if let Some(stream) =
+                        try_ready!(self.incoming.poll().map_err(Into::into))
+                    {
+                        stream
+                    } else {
+                        return Ok(Async::Ready(None));
+                    };
+
+                    if let Some(acceptor) = &mut self.acceptor {
+                        let fut = acceptor.accept(stream);
+
+                        self.state = MaybeTlsIncomingState::Accepting(fut);
+                        continue;
+                    } else {
+                        return Ok(Async::Ready(Some(MaybeTlsStream::Raw(stream))));
+                    }
+                }
+
+                MaybeTlsIncomingState::Accepting(fut) => {
+                    let stream = try_ready!(fut.poll().map_err(Box::new));
```

Comment:
Does this error provide enough context to be able to parse the error, or would it be helpful to wrap it with a (snafu) context?